_drafts/wsl2-as-a-firmware-development-environment.md

```diff
@@ -0,0 +1,485 @@
+
+> <sup>1</sup> VMWare Workstation Pro, when run headless, has the typical "X forwarding" setup
+> curve.  You can choose to emulate the graphics instead of forwarding
+> if you are OK with the overhead and latency.
+> 
+> <sup>2</sup> Not supported by default, discussed [below](#ble).
+> 
+> <sup>3</sup> % of bare metal Ubuntu 24.04 performance, based on a Zephyr benchmark, discussed [below](#performance).
+> 
+> <sup>4</sup> VMWare Workstation Pro 17 tested very poorly, but I suspect that 
+> it is because I was unable to get it to use AMD-V.  Of course, AMD-V not working
+> while WSL2 is enabled is another negative for VMWare here!  Nevertheless, I
+> suspect that it would be similar to WSL2 performance if AMD-V / Intel VT-x are working,
+> so I gave it "up to ~88%".
+> 
+> <sup>5</sup> The author's personal opinion, discussed [below](#developer-experience).
+
+### Graphics
+
+If you've previously setup X forwarding in WSL2, you'll be happy to learn that this is no longer
+necessary.  Linux GUIs "Just Work" in WSL2!
+
+### Network
+
+It’s no surprise that WSL2 can do anything you need over the network, since one of its
+primary use cases is web development.
+
+### USB
+
+[usbipd-win](https://github.com/dorssel/usbipd-win) is
+[officially recommended by Microsoft](https://learn.microsoft.com/en-us/windows/wsl/connect-usb)
+and provides a stable foundation for USB forwarding including automatic attach
+by both port and device.
+
+### BLE
+
+Even laptop BLE chips use USB, so you can easily forward an integrated or
+dongled BLE device to WSL2.  However, the default WSL2 kernel does not include
+`CONFIG_BT=y` and other modules that may be required for your specific chip. You
+will need to compile your own WSL2 kernel (only takes about 10 minutes!) and see
+the [discussion](https://github.com/dorssel/usbipd-win/discussions/310) over
+at usbipd-win.  Please [let Microsoft know](https://github.com/microsoft/WSL/issues/242)
+that this feature would be appreciated!
+
+### Performance
+
+There is a lot of confusion about WSL2 performance that's caused
+by experiences with earlier versions of WSL or use of the Windows filesystem.  Regarding
+the filesystem, it's important to understand that WSL2 creates a complete Linux
+filesystem, from `/` up, and that you want to do your work there, probably under
+`/home/<user/`, just like you would on Linux—because you are on Linux!  While it's nice
+that you can use `bash` (or `zsh`, or whatever) and Linux apps from anywhere on your Windows
+filesystem, there is a performance cost for doing so.
+
+The performance differences between WSL2 and bare metal
+Linux are [small](https://www.phoronix.com/review/windows11-wsl2-zen4/5),
+[will continue shrinking](https://www.phoronix.com/news/Microsoft-WSL2-Linux-6.6-Kernel)
+, and are not likely to meaningfully impact your workflow.
+
+## Zephyr Benchmark
+
+<div class="mermaid">
```

Comment:
This mermaid chart is great! What do you think of moving it to the end of the section? Without context, seeing this first raises a bunch of questions.

Reply:
Agreed!